### ISSUE TITLE
Refactor LSST-specific parts to use Task and Config framework

### ISSUE THREAD

**Issue by arunkannawadi** [esheldon/metadetect#220]
In the near medium term, we (Jim Bosch and I) would like to refactor the lsst modules of this package to use the Task and Config framework. This is entirely a modification to the interfaces and will still be compatible with simulations. Because the modification is not an algorithmic one, we should be able to ensure that the results are identical before and after the refactoring.

In the end, we want to have achieved the following:

- [ ] Convert the entry point (`run_metadetect` function) into a run method of a new `MetadetectTask`
- [ ] Capture all the knobs and default values (listed in [defaults.py](https://github.com/esheldon/metadetect/blob/lsst-dev/metadetect/lsst/defaults.py)) in the `ConfigClass` of the above mentioned `MetadetectTask`
- [ ] No initialization of config instances within the code. Convert all the `Task`s called internally as sub tasks of `MetadetectTask`

While neither of this is blocking a metadetection `PipelineTask`, these changes need to happen before we want to run this in production for Rubin. I am happy to start with some of these conversions but will need help from others to complete them. The upcoming DESC sprint week would be a great time to finish this refactoring.

**Comment by beckermr** [comment]
I thought there was a metadetect task that was supposed to do this. Why does you want that code here?

**Comment by esheldon** [comment]
We developed this code specifically to be separate from the tasks.

It is by design that it does not provide a Task.

This is similar to how Piff is developed independently from DM and instead there is a wrapper task for Piff, which is developed and maintained separately.

This is so that we can develop this code as independently as possible from DM processing framework, and also DM can develop their wrappers as independently from us as possible.

If there is something about the structure of our code that could be adapted to better facilitate the wrapper Task we should definitely work on that.

**Comment by arunkannawadi** [comment]
I think you are conflating `Task` with `PipelineTask` that I had hoped I clarified during the shear commissioning call two weeks ago. Refactoring parts of it into `Task`s would not limit this to running only within the DM framework. Instead, it would enable running it smoothly within that framework while maintaining the capability of running it on on-the-fly simulations and images without a butler. Not doing such a refactor upstream (i.e., here) but only in the DM fork would make pulling bug fixes and improvements (such as those from Conghao recently) into the DM fork much much harder.

Also, `metadetect` isn't comparable to `piff`, since we use `piff` as a third-party library but treat `metadetect`(https://github.com/lsst-dm/metadetect) as part of the Science Pipelines itself.

**Comment by esheldon** [comment]
I specifically said from the beginning, in like 2019, that metadetect is a third party library and should be considered so.  This has always been the intention.   metadetect is not specific to lsst, we use it for DES as well as other research projects.

In any case if you want to create a PR adding a Task to metadetect.lsst as convenience for DM would be fine.   But it should be built upon the functions and data structures we already have, which are in daily use and don't need any refactoring.

**Comment by arunkannawadi** [comment]
It's tricky to treat this as third-party since changes to Science Pipelines would affect this code, specifically `metadetect.lsst`. For the same reason, the `PipelineTask` wouldn't just be a thin wrapper around this code. I'll get a PR up soon, so it's easier to talk about some concrete changes than in abstract.

The changes would leave the generic parts of the algorithm, such as the DES aspects untouched.

**Comment by beckermr** [comment]
> While neither of this is blocking a metadetection PipelineTask, these changes need to happen before we want to run this in production for Rubin. 

I don't follow this statement precisely. Can you expand on this?

> The issue is expressing the need to expose them to the PipelineTask.

Maybe you can expand on this too?

**Comment by arunkannawadi** [comment]
Could you have a look (maybe another look?) at the WL SciUnit call (slides/recording) before I expand on them?

**Comment by beckermr** [comment]
Ah I did not realize this had been discussed. I looked over the slides but did not yet get a chance to watch the recording. Will do.

**Comment by arunkannawadi** [comment]
Well, TBH, it was not specific to `metadetection` but the material lays the groundwork for what we are proposing here.

**Comment by esheldon** [comment]
I'd like to propose a compromise.

Making this change breaks the API and thus breaks a number of codes that use this the metadetect library.

As a compromise, I propose that you implement the Task and Task Config you specify and I will write the translation later that calls to our API.

For example you will use your task config data structure and I will just write the few lines of code required to convert that into a dict that can be sent to our API.

For the actual code I will do the same, it is only a few lines of code I expect.

**Comment by arunkannawadi** [comment]
I wasn't intending this to be API-breaking from the beginning. The proposal is to move the implementation into a `Task` and have the public functions wrap the `Task` instead of other way around.

**Comment by esheldon** [comment]
There was a big misunderstanding.  I did think you were breaking the API.

Your proposal is certainly better than breaking the API, but it seems like a lot more work than the reverse that I proposed.

**Comment by arunkannawadi** [comment]
I see how I contributed to the misunderstanding; in my original post, when I meant modification to the interfaces, I meant non-public ones within this package and how it interfaces with the DM stack; certainly not the way external code is meant to interact with `metadetect`.

I'll take another look at this codebase with what you propose in mind. IIRC, what you proposed does not address my concern about exposing Configs that are being called and set internally.

**Comment by beckermr** [comment]
To be honest, I'd much rather us focus on finishing the cell-coadd work first. We have a working implementation, which is fantastic, but, as far as I know, we are likely missing a fair number of other items.

They could include:

- building of masks in healsparse for survey footprint
- selection of input visit images
- allowances for some missing pixels on the edges of cells
- finalized data structures for the cells
- implementation of Erin's new proposed interpolation scheme

What is the status of these other things?

**Comment by arunkannawadi** [comment]
Thanks, Matt. This is a good list to refer to.

- selection of input visit images

The pre-selection based on PSF and WCS quality, as well as a mask of artifacts by comparing across visits will be made before cell-based coadds are produced. The code to do that is already in place.

- allowances for some missing pixels on the edges of cells

This is already done on a development branch and Miranda is looking into this to make sure that this keeps those edges still in the buffer region.

- implementation of Erin's new proposed interpolation scheme

I haven't seen anything concrete on this and this is not on my radar.

- building of masks in healsparse for survey footprint

Miranda has put together the code for this in her [tech note draft](https://sitcomtn-137.lsst.io/v/u-mirarenee-init/index.html) that she circulated. We will get it 

- finalized data structures for the cells

This involves players at a much larger scale, and this is on Jim's plate. There is always the backup option of running an afterburner to convert it to a `CompImageHDU` that can be compressed. 

Regardless of the status on those, I'm afraid this refactoring is something that we'll have to do regardless. I see the last two points as postprocessing, which still puts this refactor ahead in terms of the timeline. This might look like code churn now, but this will save the entire Rubin production team a lot of headache, not just for commissioning but with subsequent data releases as well.

**Comment by esheldon** [comment]
The new proposal for dealing with interpolation in the coadds is very simple to implement.

Don't include interpolated pixels in the coadd.

The exception is when the pixels are interpolated in all the input images.  In that case, use the interpolated pixels.

In my version I did two passes to check for the latter case.

https://github.com/esheldon/mdet-lsst-sim/blob/12028784924fde45254000e937ff82a8d561300a/mdet_lsst_sim/coadd_fill.py
